Warning ZK-Lambda-Class Event starting (Crystal invasion event)

Ghilker opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This PR adds the crystal invasion event, a new event involving the Supermatter and the monsters from within. 
When the event starts there will be a message from centcomm announcing it, then the supermatter explode leaving a destabilized crystal that emits radiations and harmful gases; after a bit portals will spawn around the station, that will produce a number of monsters each. there are 4 types of waves and 4 types of portals
The waves are: Small, Medium, Big, Huge (each have different kind of portals that can spawn and different amounts too)
The portals are the same types of the waves, they differ from each other for the number of monsters that can spawn and the kind of monsters that can spawn (bigger portals spawn stronger monsters)
To end the event the players should stabilize the crystal by destroying the portals (for now are indestructible and they are disabled by using an anomaly neutralizer, might change that) and collecting anomaly cores; then those cores are to be put in the crystal stabilizer, an item unlockable in the tech tree. After this just inject the destabilized crystal with it and the remaining portals will close on their own (the spawned monsters will still remain tho so you have to slay them)

All the numbers are mostly eyeballed and could change if requested/with feedbacks
Code Docs will come soon!
### Warning i'm not a spriter and i can't draw for shit, so the monsters are shit, if someone want to change those in the future or have some neat monsters to give, i'll be a happy man

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
New event and crew content is always good
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Added Crystal Invasion Event (with all the mechanics, mobs and item related)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Ghilker committed:
Warning ZK-Lambda-ClassEvent starting

Ghilker committed:
added some docs

MrDoomBringer commented:
https://github.com/tgstation/tgstation/pull/50557
![image](https://user-images.githubusercontent.com/29008542/82259338-64149200-9929-11ea-9dd1-f876418d2136.png)

You could probably take a look at LemonInTheDark's implementation to get a feel for good ways to steer this PR

optimumtact commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,439 @@
+	weight = 4
+	min_players = 15
+	max_occurrences = 3
+	earliest_start = 25 MINUTES
```
lmao you're funny

Ghilker replied:
Why what's wrong with those?

ATH1909 replied:
this can happen THREE TIMES IN A SINGLE ROUND

optimumtact commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,439 @@
+	weight = 9
+	min_players = 35
+	max_occurrences = 3
+	earliest_start = 35 MINUTES
```
lmao

optimumtact commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,439 @@
+	weight = 6
+	min_players = 25
+	max_occurrences = 3
+	earliest_start = 35 MINUTES
```
lmao

optimumtact commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,439 @@
+	weight = 4
+	min_players = 35
+	max_occurrences = 3
+	earliest_start = 45 MINUTES
```
lmao

Ghilker commented:
@MrDoomBringer Potato PR is centered on making the event happen on purpose by modifying the gas composition in the SM chamber, also he used the portal storm event wich is not ideal for mine. And last one is that its event won't happen at all ever

Ghilker commented:
and @optimumtact please tell me what's wrong with those numbers, don't just laugh at them

IndieanaJones commented:
Well, for starters, you made each version of this its own event, so its possible for multiple versions of this event to trigger at once.  Along with that, having several of the same event increases the chances it'll get picked by a lot.  Along with that, you gave them pretty high weights, so the chances that they'll occur is pretty likely.  Along with that, you set each of their max occurances to 3, meaning each event can handle happen 3 times.  In total, this would be a disaster.  Your solution would be to rewrite it so its one event of varying power which can only happen once, with a rare weight.

Ghilker commented:
ahh finally, thanks @IndieanaJones i'll work to see how to fix that

wesoda25 commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,439 @@
+	///If not active the crystal will not emit radiations and gases
+	var/active = TRUE
+
+/obj/machinery/destabilized_crystal/Initialize()
+	. = ..()
+	SSshuttle.registerHostileEnvironment(src)
```
It seems a bit cruel to keep people from leaving, especially if this event occurs late game and a lot of people are dead for whatever reason.

Ghilker replied:
yeah i'll remove it in the next commit

Ghilker replied:
even tho it should stay... i mean there is a risk of universe destruction (lore only eh)

wesoda25 commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,439 @@
+	var/obj/machinery/power/supermatter_crystal/crystal = pick(sm_crystal)
+	crystal.destabilize()
+
+	priority_announce("WARNING - Detected numerous energy fluctuation generated from your Supermatter; we estimate a [wave_name] of crystal-like creatures \
+						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Collect the \[REDACTED] \
+						anomaly from the remains of the portals and use it in a crystal stabilizer to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario", "Alert", 'sound/misc/notice1.ogg')
```
```suggestion
	priority_announce("WARNING - Numerous energy fluctuations have been detected from your Supermatter; we estimate a [wave_name] of crystalline creatures \
						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Harvest a \[REDACTED] \
						anomaly from a portal, place it inside a crystal stabilizer, and inject it into your Supermatter to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario from occurring.", "Alert", 'sound/misc/notice1.ogg')
```
also wtf is a ZK-Lambda-Class Cosmic Fragmentation Scenario

Ghilker replied:
Thanks for the corrections 

is a SCP reference

Ghilker committed:
better event call, lower weight, higher pop and time requirements

Ghilker commented:
This last commit should make the event be able to be called only once per round, the min pop is 35 and the min time is 45 minutes, also i moved a bit the weight of each wave to make the more dangerous ones less likely

Shadowflame909 commented:
Do I love late-round events to spice up endgame.

Xenos kinda get old

LemonInTheDark commented:
> You could probably take a look at LemonInTheDark's implementation to get a feel for good ways to steer this PR

Bruh

LemonInTheDark commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+			kill()
+
+	sleep(10 SECONDS)
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(QDELETED(temp))
+			WARNING("No engine found.")
+			kill()
+		if(istype(temp, /obj/machinery/power/supermatter_crystal/shard))
+			message_admins("found a shard, skipping it")
+			continue
+		sm_crystal += temp
+	if(sm_crystal == null)
+		WARNING("No engine found.")
+		kill()
+	var/obj/machinery/power/supermatter_crystal/crystal = pick(sm_crystal)
```
If what you want here is the main sm crystal, use `GLOB.main_supermatter_engine`
If not, just break the loop when you find a crystal you want.

Ghilker replied:
this is made in case there are multiple crystals (like spawned by admins), it will add all them in a list and then pick one of them

LemonInTheDark replied:
Aight, that makes sense

LemonInTheDark commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Harvest a \[REDACTED] \
+						anomaly from a portal, place it inside a crystal stabilizer, and inject it into your Supermatter to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario from occurring.", "Alert", 'sound/misc/notice1.ogg')
+
+	sleep(10 SECONDS)
```
Callbacks, maybe?

LemonInTheDark commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+	sleep(10 SECONDS)
+
+	for(var/i = 0, i< portal_numbers, i++)
+		spawn_portal(wave_type)
```
Assuming you change the global list, this would be `spawn_portal(GLOB.waves[wave_name])`

Ghilker replied:
done

LemonInTheDark commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+/obj/machinery/destabilized_crystal/Destroy()
+	return..()
+
+/obj/machinery/destabilized_crystal/process()
+	if(active)
+		if(prob(45))
+			src.fire_nuclear_particle()
+			radiation_pulse(src, 250, 6)
+		var/turf/T = loc
+		var/datum/gas_mixture/env = T.return_air()
+		var/datum/gas_mixture/removed
+		var/gasefficency = 0.5
+		removed = env.remove(gasefficency * env.total_moles())
+		removed.assert_gases(/datum/gas/bz, /datum/gas/stimulum, /datum/gas/nitryl, /datum/gas/miasma)
+		removed.gases[/datum/gas/bz][MOLES] += 5.5
+		removed.gases[/datum/gas/stimulum][MOLES] += 4.5
+		removed.gases[/datum/gas/nitryl][MOLES] += 6.75
+		removed.gases[/datum/gas/miasma][MOLES] += 10.5
+		env.merge(removed)
+		air_update_turf()
```
How slow is this in testing/how large can this get?

LemonInTheDark replied:
I could see this making a lot of active atmos turfs

Ghilker replied:
the max this got was filling up the SM chamber with gases, it was pretty slow but i could remove some of the gases (like stim and nitryl) to remove some of the load to the atmos

LemonInTheDark replied:
Might be a good idea in general, depending on how you want to limit access to those gasses.

Ghilker replied:
removed stim and nitryl productions

LemonInTheDark commented on code/modules/power/supermatter/supermatter.dm:
```diff
@@ -1124,6 +1124,13 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 		for(var/j in 1 to zap_count)
 			supermatter_zap(target, new_range, zap_str, targets_hit, zap_flags)
 
+/obj/machinery/power/supermatter_crystal/proc/destabilize()
+	var/turf/T = get_turf(src)
+	if(T) //If something fucks up we blow anyhow. This fix is 4 years old and none ever said why it's here. help.
+		qdel(src)
+		new/obj/machinery/destabilized_crystal(T)
+		explosion(T,0,4,15,1,1,1)
```
Minor thing, removing the comment here should be fine, the comment isn't really accurate in this case either.

Ghilker replied:
removed

LemonInTheDark reviewed (commented):
I'm a little unsure about how this would interact with engi players well playing, so some more detail would be nice, however this seems interesting, and interesting things have their place.

I've left a couple of nitpicks about how things are stored, and one or two concerns over speed. Take a read when you get the chance.

Ghilker commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
+
+	sleep(10 SECONDS)
```
how does callback works?

LemonInTheDark replied:
Looks like this `addtimer(CALLBACK(object_to_call_on, .proc_you_want_to_call, arguments), time in deciseconds)`
Instead of relying on byond to do the sleep/recall for us, it uses the timer system we've built, which tends to be a bit safer on the overall. Not sure if this is a good usecase or not, but it should help prevent unintended issues in future.

Ghilker replied:
done

Ghilker commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+GLOBAL_LIST_INIT(small_wave, list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=1)) //5
+GLOBAL_LIST_INIT(medium_wave, list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=1)) //8
+GLOBAL_LIST_INIT(big_wave, list(/obj/structure/crystal_portal/small=5, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=2, /obj/structure/crystal_portal/huge=1)) //11
+GLOBAL_LIST_INIT(huge_wave, list(/obj/structure/crystal_portal/small=7, /obj/structure/crystal_portal/medium=5, /obj/structure/crystal_portal/big=3, /obj/structure/crystal_portal/huge=2)) //17
```
done next commit

Ghilker commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,408 @@
+			"medium wave" = 30,
+			"big wave" = 7,
+			"huge wave" = 3))
+	switch(wave_name)
+		if("small wave")
+			wave_type = GLOB.small_wave
+			portal_numbers = pick(2, 3, 4, 5)
+		if("medium wave")
+			wave_type = GLOB.medium_wave
+			portal_numbers = pick(4, 5, 6, 7)
+		if("big wave")
+			wave_type = GLOB.big_wave
+			portal_numbers = pick(6, 7, 8, 9)
+		if("huge wave")
+			wave_type = GLOB.huge_wave
+			portal_numbers = pick(9, 10, 11, 12)
+		else
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
```
done next commit

Ghilker committed:
even better event calls, requested changes, fixes and stuff

Ghilker committed:
more changes and additions

OverlyGeneric commented:
If you wanted to, you could use the old lavaland mobs as placeholders.

Hivelords, old goliaths, and basilisks are hardly seen.

LemonInTheDark commented:
Ok this is getting close to plagiarism

ATH1909 commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,412 @@
+This section is for the destabilized SM
+*/
+
+/obj/machinery/destabilized_crystal
```
why is this its own separate object instead of a variable or something for the SM crystal?

Ghilker replied:
because is better to have another object to handle this instead of the SM to make it more modular (we can change this obj far more easily) and to not bloat too much the SM code

ATH1909 commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,412 @@
+	icon_living = "crystal_minion"
+	icon_dead = "crystal_minion"
+	gender = NEUTER
+	mob_biotypes = MOB_UNDEAD|MOB_HUMANOID
```
shouldn't they be inorganic, not undead?

Ghilker replied:
i mean they are skeletons with crystals on them

nemvar replied:
Plasmamen are also skeletons but they are classified as inorganics. Your call though.

ATH1909 commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,412 @@
+	harm_intent_damage = 5
+	melee_damage_lower = 5
+	melee_damage_upper = 15
+/mob/living/simple_animal/hostile/crystal_monster/boss
```
these crystal dudes are kind of pathetically weak

also, why do crystal minions take half as much damage from barehanded punches than the other kinds of crystal monsters do?

Ghilker replied:
i just eyeballed the numbers, i have no clue how powerful is powerful, what would be a good reference point to do this?

Ghilker replied:
i checked the spiders, made the crystal monster around those values

Ghilker committed:
more stuff

Ghilker committed:
20 minutes not 2 minutes

Ghilker commented:
more stuff, added the Destabilized crystal explosion if is not closed in time. The event also spawns randomly explosive flux anomalies (with smaller explosion than the normal ones). Plus there is a second wave of portals after 10 minutes, after another 10 minutes the crystal will collapse generating more portals and a big explosion (might bump up the numbers to 15 minutes).
Changed the amount of damages that the crystal monsters deal

kriskog reviewed (dismissed):
.

Ghilker committed:
Merge branch 'master' into ZK-Lambda-ClassEvent
Merge master

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+
+	if(!spawners.len)
+		message_admins("No landmarks on the station, aborting")
+		return MAP_ERROR
```
returning MAP_ERROR here seems really odd, the return isn't even used and it is a pretty unrelated define.

Ghilker replied:
i was following the meteor event and other events, where it should be a map error if the landmarks are not present to spawn the things. I could just remove it and do a normal return

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	for(var/obj/effect/landmark/event_spawn/temp in GLOB.generic_event_spawns)
+		if(QDELETED(temp))
+			continue
+		if(is_station_level(temp.loc.z))
```
just do ``temp.z``, it will always be the same as the z of the location.

Ghilker replied:
aight

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+
+	if(!spawners.len)
+		message_admins("No landmarks on the station, aborting")
+		return MAP_ERROR
```
Again, MAP_ERROR doesn't look correct here. Just return nothing.

Ghilker replied:
ditto

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	if(!spawners.len)
+		message_admins("No landmarks on the station, aborting")
+		return MAP_ERROR
+	var/obj/spawner = pick_n_take(spawners)
```
This pick_n_take does not work as you probably intended it to since the spawners list gets calculated seperately for each portal.

Ghilker replied:
so just pick?

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	priority_announce("WARNING - The crystal has reached critical instability point. ZK-Event inbound, please do not panic, anyone who panics will \
+						be terminated on the spot. Have a nice day", "Alert")
+	sound_to_playing_players('sound/machines/alarm.ogg')
+	sleep(15 SECONDS)
```
Replace this with a timer.

Ghilker replied:
can `addtimer` be used without a callback function? because i didn't want to make it into yet another proc

Rohesie replied:
It cannot, but it's fine to add another proc. Just add `do_zk_event()`

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+/obj/machinery/destabilized_crystal/process()
+	if(active)
+		if(prob(75))
+			src.fire_nuclear_particle()
```
Might have changed but last I checked nuclear particles were extremely dangerous, as in, get his once and spent 20 minutes chugging anti rad meds even though you wore a radsuit.
This might make it basically impossible to actually fix the crystal without dying since you need to stand next to it for more than 15 seconds.

Ghilker replied:
i'll remove them, is already difficult to fix

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	qdel(src)
+
+/obj/machinery/destabilized_crystal/attackby(obj/item/W, mob/living/user, params)
+	if(!istype(W) || (W.item_flags & ABSTRACT) || !istype(user))
```
Typecheck on the item and check for the item flag are not needed.
If someone somehow gets an abstract stabilizer, they deserve to fix this machine.

Ghilker replied:
aight, that would be a funny meme

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	var/filled = FALSE
+
+/obj/item/crystal_stabilizer/attackby(obj/item/W, mob/living/user, params)
+	if(!istype(W) || (W.item_flags & ABSTRACT) || !istype(user))
```
should call parent.

Ghilker replied:
ok

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	slot_flags = ITEM_SLOT_BACK
+	w_class = WEIGHT_CLASS_HUGE
+	///The stabilizer is one use only
+	var/filled = FALSE
```
maybe tell people that it is loaded over an examine message.

Ghilker replied:
will do

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+		return
+	if(istype(W, /obj/item/anomaly_neutralizer))
+		to_chat(user, "<span class='notice'>You start closing the [src]</span>")
+		if(W.use_tool(src, user, 5.5 SECONDS, volume=100))
```
what does the volume parameter do?

Ghilker replied:
it should be for any sound effects

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	mob_types = list(/mob/living/simple_animal/hostile/crystal_monster/minion, /mob/living/simple_animal/hostile/crystal_monster/thug)
+	spawn_text = "emerges from"
+	faction = list("crystal")
+/obj/structure/crystal_portal/medium
```
please add some lines between the datums, also, remove the redundant variables.
For example, ``spawn_text`` or ``icon`` are the same for all of these so there is no need to set them for the child types.

Ghilker replied:
sure

nemvar commented on code/modules/power/supermatter/supermatter.dm:
```diff
@@ -1124,6 +1124,13 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
+	if(T)
+		qdel(src)
+		new/obj/machinery/destabilized_crystal(T)
+		explosion(T,round(portal_numbers/10),round(portal_numbers/5),round(portal_numbers),1,1,1)
```
This should probably not have an devestation tiles at all, considering how that completely negates the gas generation.

Ghilker replied:
true that, i'll set the devastation to zero

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+GLOBAL_LIST_INIT(waves, list("small wave" = list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=1),
```
I'm not convinced these actually need to be globals but this name is not good. Should at the very least be renamed to ``crystal_invasion_waves"

Ghilker replied:
done

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(QDELETED(temp))
```
There shouldn't even be qdeleted machines in GLOB.machines and even if there are, just skip them instead of canceling the event (which the kill() call doesn't even do correctly).

Ghilker replied:
better safe than sorry, if an admin delete the SM crystal or something weird happened i'll just skip ahead since there is a kill() proc later if there is no crystal around

Rohesie replied:
Don't check for qdeleted there at all.
If this happens we definitely want things to break.
Instead, make sure it doesn't happen.

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+				to_chat(user, "<span class='notice'>The [src] is reacting violently with the fluid!</span>")
+				src.fire_nuclear_particle()
+				radiation_pulse(src, 2500, 6)
+				if(W.use_tool(src, user, 7.5 SECONDS, volume=100))
```
Please turn these into early returns.

Ghilker replied:
uh how? is a three processes one after another, i don't see a way of making it early return

nemvar commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+			if("Big Portal")
+				explosion(loc, 1,3,5)
+			if("Huge Portal")
+				explosion(loc, 2,5,7)
```
Fixing the crystal before closing all the portals will cause them to explode. Is that intentional?

Ghilker replied:
yes

Rohesie replied:
This will happen before the crystal portal is moved to nullspace, meaning the explosion will affect it.
Avoid adding these interactions with the outside world on `Destroy()` if you can avoid it, instead make a precursor proc that calls for its deletion.

Ghilker commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+			"huge wave" = 3))
+	switch(wave_name)
+		if("small wave")
+			portal_numbers = pick(2, 3, 4, 5)
```
done in next commit

Ghilker commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+			continue
+		sm_crystal += temp
+	if(sm_crystal == null)
+		WARNING("No engine found.")
```
the admin message was a leftover, thanks for that.
just log_game?

Ghilker commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,477 @@
+	addtimer(CALLBACK(src, .proc/spawn_portals), 10 SECONDS)
+
+/datum/round_event/crystal_invasion/proc/spawn_portals()
+	for(var/i = 0, i< portal_numbers, i++)
```
next commit

Ghilker committed:
requested changes and mob flag turned to MINERAL

Ghilker committed:
forgot this

Ghilker commented:
stale bot go away

LemonInTheDark commented:
I believe

Rohesie commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -147,6 +147,7 @@
 	var/canshock = FALSE
 	var/shockdamage = 20
 	var/explosive = TRUE
+	var/is_zk = FALSE
```
Please document this variable and give it a more informative name.
What the hell is zk?

Rohesie commented on code/modules/power/supermatter/supermatter.dm:
```diff
@@ -1104,6 +1104,13 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
+/obj/machinery/power/supermatter_crystal/proc/destabilize(portal_numbers)
+	var/turf/T = get_turf(src)
+	if(T)
+		qdel(src)
+		new/obj/machinery/destabilized_crystal(T)
+		explosion(T,0,round(portal_numbers/5),round(portal_numbers),1,1,1)
```
Code after `qdel(src)` may not be executed, only `qdel` at the end of it.
Also, under which circumstances could you call this proc with the SM being in nullspace?

Rohesie commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,465 @@
+			to_chat(user, "<span class='notice'>\The [W] is empty!</span>")
+			return
+		to_chat(user, "<span class='notice'>You carefully begin inject \the [src] with \the [W]... please don't move untill all the steps are finished</span>")
+		if(W.use_tool(src, user, 5 SECONDS, volume=100))
+			to_chat(user, "<span class='notice'>Seems that \the [src] internal resonance is fading with the fluid!</span>")
+			playsound(get_turf(src), 'sound/effects/supermatter.ogg', 35, TRUE)
+			if(W.use_tool(src, user, 6.5 SECONDS, volume=100))
+				to_chat(user, "<span class='notice'>The [src] is reacting violently with the fluid!</span>")
+				src.fire_nuclear_particle()
+				radiation_pulse(src, 2500, 6)
+				if(W.use_tool(src, user, 7.5 SECONDS, volume=100))
```
```suggestion
		if(!W.use_tool(src, user, 5 SECONDS, volume=100))
			return
		to_chat(user, "<span class='notice'>Seems that \the [src] internal resonance is fading with the fluid!</span>")
		playsound(get_turf(src), 'sound/effects/supermatter.ogg', 35, TRUE)
		if(!W.use_tool(src, user, 6.5 SECONDS, volume=100))
			return
		to_chat(user, "<span class='notice'>The [src] is reacting violently with the fluid!</span>")
		src.fire_nuclear_particle()
		radiation_pulse(src, 2500, 6)
		if(!W.use_tool(src, user, 7.5 SECONDS, volume=100))
			return
```

Rohesie commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,465 @@
+
+/obj/item/crystal_stabilizer/attackby(obj/item/W, mob/living/user, params)
+	. = ..()
+	if(!istype(W) || (W.item_flags & ABSTRACT) || !istype(user))
```
```suggestion
	if((W.item_flags & ABSTRACT) || !istype(user))
```
`attackby()` should never be triggered by a non-item

Rohesie commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,465 @@
+/obj/structure/crystal_portal/attack_animal(mob/living/simple_animal/M)
+	if(faction_check(faction, M.faction, FALSE)&&!M.client)
+		return
+	..()
```
```suggestion
	return ..()
```

Rohesie commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,465 @@
+	return ..()
+
+/obj/structure/crystal_portal/attack_animal(mob/living/simple_animal/M)
+	if(faction_check(faction, M.faction, FALSE)&&!M.client)
```
```suggestion
	if(faction_check(faction, M.faction, FALSE) && !M.client)
```
No need to save up on spaces. They make things more clear.

Rohesie commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,465 @@
+		return
+	if(istype(W, /obj/item/anomaly_neutralizer))
+		to_chat(user, "<span class='notice'>You start closing the [src]</span>")
+		if(W.use_tool(src, user, 5.5 SECONDS, volume=100))
```
```suggestion
		if(W.use_tool(src, user, 5.5 SECONDS, volume = 100))
```
Spaces are very cheap, try to use them for clarity.

Rohesie commented on code/modules/events/crystal_event.dm:
```diff
@@ -0,0 +1,465 @@
+	color = "#F97575"
+	max_mobs = 12
+	spawn_time = 140
+	mob_types = list(/mob/living/simple_animal/hostile/crystal_monster/minion, /mob/living/simple_animal/hostile/crystal_monster/thug,\
+					 /mob/living/simple_animal/hostile/crystal_monster/recruit, /mob/living/simple_animal/hostile/crystal_monster/killer, \
+					 /mob/living/simple_animal/hostile/crystal_monster/boss)
```
```suggestion
	mob_types = list(/mob/living/simple_animal/hostile/crystal_monster/minion,
		/mob/living/simple_animal/hostile/crystal_monster/thug,
		/mob/living/simple_animal/hostile/crystal_monster/recruit,
		/mob/living/simple_animal/hostile/crystal_monster/killer,
		/mob/living/simple_animal/hostile/crystal_monster/boss,
		)
```
You don't need `\` unless if you are formatting text really.
Adding a comma on the last list member won't add a null entry and will diminish lines changed if a new one is introduced later.
This pattern is a bit more clear to me, but it's optional.

Rohesie commented:
Do the changes if you plan to re-open the PR. Closing it due to staleness.
